internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+	}
+}
+
+// checkTTU looks up all tuples of the target tupleset relation on the provided object and for each one
+// of them evaluates the computed userset of the TTU rewrite rule for them.
+func (c *ConcurrentChecker) checkTTU(parentctx context.Context, req *dispatcher.DispatchCheckRequest, rewrite *openfgapb.Userset) CheckHandlerFunc {
+
+	typesys, ok := typesystem.TypesystemFromContext(parentctx)
+	if !ok {
+		panic("typesystem missing in context")
+	}
+
+	return func(ctx context.Context) (*openfgapb.CheckResponse, error) {
+
+		ctx = typesystem.ContextWithTypesystem(ctx, typesys)
+
+		tuplesetRelation := rewrite.GetTupleToUserset().GetTupleset().GetRelation()
+		computedRelation := rewrite.GetTupleToUserset().GetComputedUserset().GetRelation()
+
+		tk := req.GetTupleKey()
+
+		iter, err := c.ds.Read(
+			ctx,
+			req.GetStoreID(),
+			tuple.NewTupleKey(tk.GetObject(), tuplesetRelation, ""),
+		)
+		if err != nil {
+			return &openfgapb.CheckResponse{Allowed: false}, err
+		}
+		defer iter.Stop()
+
+		// filter out invalid tuples yielded by the database iterator
+		filteredIter := storage.NewFilteredTupleKeyIterator(
+			storage.NewTupleKeyIteratorFromTupleIterator(iter),
+			validation.FilterInvalidTuples(typesys),
+		)
+		defer filteredIter.Stop()
+
+		var handlers []CheckHandlerFunc
+		for {
+			t, err := filteredIter.Next(ctx)
+			if err != nil {
+				if err == storage.ErrIteratorDone {
+					break
+				}
+
+				return &openfgapb.CheckResponse{Allowed: false}, err
+			}
+
+			userObj, _ := tuple.SplitObjectRelation(t.GetUser())
+
+			tupleKey := &openfgapb.TupleKey{
+				Object:   userObj,
+				Relation: computedRelation,
+				User:     tk.GetUser(),
+			}
+
+			handlers = append(handlers, c.dispatch(
+				ctx,
+				&dispatcher.DispatchCheckRequest{
+					StoreID:              req.GetStoreID(),
+					AuthorizationModelID: req.GetAuthorizationModelID(),
+					TupleKey:             tupleKey,
+					ResolutionMetadata: &dispatcher.ResolutionMetadata{
+						Depth: req.GetResolutionMetadata().Depth - 1,
+					},
+				}))
+		}
+
+		if len(handlers) > 0 {
+			return union(ctx, c.concurrencyLimit, handlers...)
+		}
+
+		return &openfgapb.CheckResponse{Allowed: false}, nil
+	}
+}
+
+func (c *ConcurrentChecker) checkSetOperation(
+	ctx context.Context,
+	req *dispatcher.DispatchCheckRequest,
+	setOpType setOperatorType,
+	reducer CheckFuncReducer,
+	children ...*openfgapb.Userset,
+) CheckHandlerFunc {
+
+	var handlers []CheckHandlerFunc
+
+	switch setOpType {
+	case unionSetOperator, intersectionSetOperator:
+		for _, child := range children {
+			handlers = append(handlers, c.checkRewrite(ctx, req, child))
+		}
+	case exclusionSetOperator:
+		if len(children) != 2 {
+			panic(fmt.Sprintf("expected two rewrite operands in exclusion operator, but got '%d'", len(children)))
+		}
+
+		handlers = append(handlers,
+			c.checkRewrite(ctx, req, children[0]),
+			c.checkRewrite(ctx, req, children[1]),
+		)
+	}
```

Comment:
```suggestion
	switch setOpType {
	case unionSetOperator, intersectionSetOperator, exclusionSetOperator:
		for _, child := range children {
			handlers = append(handlers, c.checkRewrite(ctx, req, child))
		}
	default:
		// unknown operator
	}
```

and check for the right size of `children` for the exclusion case within `checkRewrite`

Reply:
Nice, I like that it's more uniform 👍  Done.